### PR TITLE
Initial commit of System.Collections partial facade

### PR DESCRIPTION
**Pull request by FiveTimesTheFun**
Initial commit of System.Collections partial facade.  This partial
facade uses the new FindPartialFacadeContractReference and
FillPartialFacade targets to create a facade against mscorlib.  We
compile against the implementation assemblies mscorlib.dll (from the
Microsoft.DotNet.CoreCLR package), System.Diagnostics.Debug.dll,
and System.Runtime.dll.

Tests continue to compile against the System.Collections nuget package
until we have the ability to reference the contract for compilation
but copy the correct implementation assembly at test time.

### CODE REVIEW

**Comment by stephentoub** [comment]
> Tests continue to compile against the System.Collections nuget package until we have the ability to reference the contract for compilation but copy the correct implementation assembly at test time.

@mellinoe appears to have worked around that in System.Runtime.Handles, assuming I'm reading the comment correctly.  Could we do the same thing here?
https://github.com/dotnet/corefx/commit/a7acef765e77e0a1dd80e3d1e56f00b74402028a#diff-4cdbf1333a93ee83e7c15b05f49eb713R18

File: src/Common/src/System/Collections/Generic/LowLevelComparer.cs [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+            if (typeof(T) == typeof(Int32))
+                return (IComparer<T>)(Object)(new LowLevelInt32Comparer());
+
+            return new LowLevelComparer<T>();
+        }
+
+        private static IComparer<T> _default;
+    }
+
+    internal class LowLevelInt32Comparer : IComparer<Int32>
```
Comment (stephentoub) [inline src/Common/src/System/Collections/Generic/LowLevelComparer.cs]: sealed?

File: src/System.Collections/src/System/Collections/Generic/Queue.cs [hunk 2]
```diff
@@ -0,0 +1,461 @@
...
+    {
+        private T[] _array;
+        private int _head;       // First valid element in the queue
+        private int _tail;       // Last valid element in the queue
+        private int _size;       // Number of elements.
+        private int _version;
+        private Object _syncRoot;
+
+        private const int _MinimumGrow = 4;
+        private const int _ShrinkThreshold = 32;
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/Queue.cs]: While it might be nice if this was used, it doesn't appear to be.

File: src/System.Collections/src/System/Collections/BitArray.cs [hunk 3]
```diff
@@ -0,0 +1,544 @@
...
+                index = -1;
+            }
+        }
+
+        private int[] m_array;
+        private int m_length;
+        private int _version;
+        private Object _syncRoot;
+
+        private const int _ShrinkThreshold = 256;
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/BitArray.cs]: Per style guidelines, constants aren't supposed to begin with underscores, right?

File: src/System.Collections/src/System/Collections/Generic/BitHelper.cs [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+    /// IsMarked -- pass in values as indices into the logical bit array, and it will be mapped
+    /// to the position within the array of ints.
+    /// 
+    /// FUTURE OPTIMIZATIONS:
+    /// A method such as FindFirstMarked/Unmarked Bit would be useful for callers that operate 
+    /// on a bit array and then need to loop over it. In particular, if it avoided visiting 
+    /// every bit, it would allow good perf improvements when the bit array is sparse.
+    /// </summary>
+    unsafe internal class BitHelper
+    {   // should not be serialized
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/BitHelper.cs]: What does this "should not be serialized" comment refer to?

File: src/System.Collections/src/System/Collections/Generic/BitHelper.cs [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+                    m_arrayPtr[bitArrayIndex] |= (MarkedBitFlag << (bitPosition % IntSize));
+                }
+            }
+            else
+            {
+                int bitArrayIndex = bitPosition / IntSize;
+                if (bitArrayIndex < m_length && bitArrayIndex >= 0)
+                {
+                    m_array[bitArrayIndex] |= (MarkedBitFlag << (bitPosition % IntSize));
+                }
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/BitHelper.cs]: In both this and the subsequent function, the only difference between the if and else block appears to be whether the result is stored into m_arrayPtr or m_array.  You could avoid duplicated code by just having the conditional around that piece rather than the whole thing.

File: src/System.Collections/src/System/Collections/Generic/Comparer.cs [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+
+using System;
+using System.Collections;
+using System.Diagnostics.Contracts;
+
+namespace System.Collections.Generic
+{
+    public abstract class Comparer<T> : IComparer, IComparer<T>
+    {
+        protected Comparer()
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/Comparer.cs]: This shouldn't be necessary. Is the ctor here just to be explicit?

File: src/System.Collections/src/System/Collections/Generic/Comparer.cs [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+        {
+        }
+
+        public static Comparer<T> Default
+        {
+            get
+            {
+                if (_default == null)
+                    _default = CreateComparer();
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/Comparer.cs]: Probably something to be done as a separate commit, but see https://github.com/dotnet/corefx/issues/780

File: src/Common/src/System/Collections/Generic/LowLevelComparer.cs [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+//
+// LowLevelComparer<T> emulates the desktop type System.Collections.Generic.Comparer<T>
+//
+
+using System;
+using System.Collections;
+
+namespace System.Collections.Generic
+{
+    internal sealed class LowLevelComparer<T> : IComparer<T>
```
Comment (stephentoub) [inline src/Common/src/System/Collections/Generic/LowLevelComparer.cs]: Why is this type necessary?  Why is just Comparer<T> (also in this assembly) insufficient?

File: src/System.Collections/src/System/Collections/Generic/Comparer.cs [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+        {
+            if (x == null) return y == null ? 0 : -1;
+            if (y == null) return 1;
+            if (x is T && y is T) return Compare((T)x, (T)y);
+            throw new ArgumentException(SR.Argument_InvalidArgumentForComparison);
+        }
+
+        private static Comparer<T> CreateComparer()
+        {
+            // NUTC compiler optimization see comments in EqualityComparer.cs
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/Comparer.cs]: See my comment on the same code in ```LowLevelComparer<T>```.

File: src/System.Collections/src/System/Collections/Generic/Comparer.cs [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+                return (Comparer<T>)(Object)(new Int32Comparer());
+
+            return new DefaultComparer<T>();
+        }
+
+        private static Comparer<T> _default;
+    }
+
+
+    internal class DefaultComparer<T> : Comparer<T>
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/Comparer.cs]: Just for hygiene reasons, it'd be nice if all of these derived types could be sealed.

File: src/System.Collections/src/System/Collections/Generic/Dictionary.cs [hunk 11]
```diff
@@ -0,0 +1,1265 @@
...
+        private int version;
+
+        private int freeList;
+        private int freeCount;
+        private IEqualityComparer<TKey> comparer;
+        private KeyCollection keys;
+        private ValueCollection values;
+        private Object _syncRoot;
+
+        // constants for serialization
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/Dictionary.cs]: Are these constants used anywhere?

File: src/System.Collections/src/System/Collections/Generic/Dictionary.cs [hunk 12]
```diff
@@ -0,0 +1,1265 @@
...
+                return true;
+            }
+            return false;
+        }
+
+        public void Clear()
+        {
+            if (count > 0)
+            {
+                for (int i = 0; i < buckets.Length; i++) buckets[i] = -1;
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/Dictionary.cs]: I forget now... did we add anything to the style guidelines around one line loops like this?

File: src/System.Collections/src/System/Collections/Generic/Dictionary.cs [hunk 13]
```diff
@@ -0,0 +1,1265 @@
...
+            if (key == null)
+            {
+                throw new ArgumentNullException("key");
+            }
+
+            if (buckets == null) Initialize(0);
+            int hashCode = comparer.GetHashCode(key) & 0x7FFFFFFF;
+            int targetBucket = hashCode % buckets.Length;
+
+#if FEATURE_RANDOMIZED_STRING_HASHING
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/Dictionary.cs]: Is this compilation constant still needed?  We don't appear to be setting it; should we?

File: src/System.Collections/src/System/Collections/Generic/EqualityComparer.cs [hunk 14]
```diff
@@ -0,0 +1,336 @@
...
+
+        public static EqualityComparer<T> Default
+        {
+            get
+            {
+                if (_default == null)
+                {
+                    object comparer;
+
+                    // NUTC compiler is able to static evalulate the conditions and only put the necessary branches in finally binary code,
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/EqualityComparer.cs]: The same is true of the JIT; at least it was in .NET 4.5.  Hopefully that hasn't changed with RyuJIT.

File: src/System.Collections/src/System/Collections/Generic/EqualityComparer.cs [hunk 15]
```diff
@@ -0,0 +1,336 @@
...
+            if (x == null || y == null)
+                return false;
+            if ((x is T) && (y is T))
+                return Equals((T)x, (T)y);
+            throw new ArgumentException(SR.Argument_InvalidArgumentForComparison);
+        }
+    }
+
+    //
+    // ProjectN compatiblity notes:
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/EqualityComparer.cs]: Is this comment still relevant?  And this should be ".NET Native" rather than "ProjectN", right?

**Comment by FiveTimesTheFun** [comment]
Thanks Stephen - I updated the test project to copy the product assembly to the output directory through a project reference and updated the initial commit.  I'm going through the rest of your comments now.

File: src/System.Collections/src/System/Collections/Generic/LinkedList.cs [hunk 16]
```diff
@@ -0,0 +1,606 @@
...
+    public class LinkedList<T> : ICollection<T>, System.Collections.ICollection, IReadOnlyCollection<T>
+    {
+        // This LinkedList is a doubly-Linked circular list.
+        internal LinkedListNode<T> head;
+        internal int count;
+        internal int version;
+        private Object _syncRoot;
+
+        // names for serialization
+        const String VersionName = "Version";
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/LinkedList.cs]: Same question about the serialization constants.

File: src/System.Collections/src/System/Collections/Generic/LinkedList.cs [hunk 17]
```diff
@@ -0,0 +1,606 @@
...
+            }
+
+            public void Dispose()
+            {
+            }
+        }
+    }
+
+    // Note following class is not serializable since we customized the serialization of LinkedList. 
+    [System.Runtime.InteropServices.ComVisible(false)]
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/LinkedList.cs]: Are such ComVisible attributes still necessary?

File: src/System.Collections/src/System/Collections/Generic/List.cs [hunk 18]
```diff
@@ -0,0 +1,1381 @@
...
+    {
+        private const int _defaultCapacity = 4;
+
+        private T[] _items;
+        [ContractPublicPropertyName("Count")]
+        private int _size;
+        private int _version;
+        private Object _syncRoot;
+
+        static readonly T[] _emptyArray = new T[0];
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/List.cs]: Style guideline naming: this should be s_emptyArray.  It should also be explicitly marked as private.

File: src/System.Collections/src/System/Collections/Generic/List.cs [hunk 19]
```diff
@@ -0,0 +1,1381 @@
...
+    // elements. A List has a capacity, which is the allocated length
+    // of the internal array. As elements are added to a List, the capacity
+    // of the List is automatically increased as required by reallocating the
+    // internal array.
+    // 
+    [DebuggerTypeProxy(typeof(ICollectionDebugView<>))]
+    [DebuggerDisplay("Count = {Count}")]
+    public class List<T> : IList<T>, System.Collections.IList, IReadOnlyList<T>
+    {
+        private const int _defaultCapacity = 4;
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/List.cs]: Style guideline naming: Constants should be PascalCased, without a leading underscore.

File: src/Common/src/System/Collections/Generic/LowLevelComparer.cs (hunk 1 above)
Comment (FiveTimesTheFun) [inline src/Common/src/System/Collections/Generic/LowLevelComparer.cs]: Thanks Stephen - I updated to make all internal classes sealed where appropriate.

File: src/System.Collections/src/System/Collections/Generic/Comparer.cs (hunk 10 above)
Comment (FiveTimesTheFun) [inline src/System.Collections/src/System/Collections/Generic/Comparer.cs]: Thanks, updated.

File: src/System.Collections/src/System/Collections/Generic/Queue.cs [hunk 20]
```diff
@@ -0,0 +1,461 @@
...
+        private const int _GrowFactor = 200;  // double each time
+        private const int _DefaultCapacity = 4;
+        static T[] _emptyArray = new T[0];
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/Queue.cs]: Same comments about static field and constant names... I'll stop commenting on these :)

File: src/System.Collections/src/System/Collections/Generic/Dictionary.cs (hunk 12 above)
Comment (FiveTimesTheFun) [inline src/System.Collections/src/System/Collections/Generic/Dictionary.cs]: I don't see anything on the Wiki:
https://github.com/dotnet/corefx/wiki/Coding-style

Willing to start an issue to discuss, though.  Would personally like to keep constructs like
if (expr) return val1;
else return val2;
but will update this one for better readability.

File: src/System.Collections/src/System/Collections/Generic/List.cs (hunk 18 above)
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/List.cs]: Consider using Array.Empty<T>() instead of these _emptyArray fields in these various types.

File: src/System.Collections/src/System/Collections/Generic/QueueDebugView.cs [hunk 21]
```diff
@@ -0,0 +1,31 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Diagnostics;
+
+namespace System.Collections.Generic
+{
+    internal sealed class QueueDebugView<T>
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/QueueDebugView.cs]: Since most of these debug views just do ToArray, could we collapse these into views based on the interfaces?

File: src/System.Collections/src/System/Collections/Generic/SortedSet.cs [hunk 22]
```diff
@@ -0,0 +1,2552 @@
...
+    public class SortedSet<T> : ISet<T>, ICollection<T>, ICollection, IReadOnlyCollection<T>
+    {
+        #region local variables/constants
+        Node root;
+        IComparer<T> comparer;
+        int count;
+        int version;
+        private Object _syncRoot;
+
+        private const String ComparerName = "Comparer";
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/SortedSet.cs]: Same question about serialization constants.  Will stop commenting on these.

File: src/System.Collections/src/System/Collections/Generic/Dictionary.cs (hunk 13 above)
Comment (FiveTimesTheFun) [inline src/System.Collections/src/System/Collections/Generic/Dictionary.cs]: Thanks Stephen - this feature is enabled for coreclr (and set for mscorlib, https://github.com/dotnet/coreclr/blob/524f5463406df065474b342086e6dfe38ddfbda3/clr.coreclr.props#L78 ) and should be enabled here.

Comment (FiveTimesTheFun) [inline src/System.Collections/src/System/Collections/Generic/Dictionary.cs]: It looks like it should also be enabled for https://github.com/dotnet/corefx/blob/7116584186f8f3a886616aaf8cb5d4a982c60e27/src/System.Collections.NonGeneric/src/System/Collections/Hashtable.cs#L821 .

It looks like our corefx common msbuild files only define "DEBUG" and "TRACE", while our buildtools msbuild files only define "DEBUGRESOURCES" and "SIGNED", and individual libraries have so far defined 'feature' constants like this one themselves.  I'll update this PR and add a new PR for System.Collections.NonGeneric.

File: src/System.Collections/src/System/Collections/Generic/SortedSet.cs [hunk 23]
```diff
@@ -0,0 +1,2552 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+// #define USING_HASH_SET
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/SortedSet.cs]: Can the #ifs for USING_HASH_SET be deleted, or should this be added to the .csproj?

File: src/System.Collections/src/project.json [hunk 24]
```diff
@@ -0,0 +1,24 @@
+{
+  "dependencies": {
+//    "System.Collections": "4.0.10-beta-22809",
```
Comment (stephentoub) [inline src/System.Collections/src/project.json]: What are all of these '//'s?

File: src/System.Collections/src/project.json [hunk 25]
```diff
@@ -0,0 +1,24 @@
...
+//    "System.Runtime.Extensions": "4.0.10-beta-22809",
+//    "System.Runtime.Handles": "4.0.0-beta-22809",
+//    "System.Runtime.InteropServices": "4.0.20-beta-22809",
+//    "System.Text.Encoding": "4.0.10-beta-22809",
+//    "System.Text.Encoding.Extensions": "4.0.10-beta-22809",
+//    "System.Threading": "4.0.10-beta-22809",
+//    "System.Threading.Tasks": "4.0.10-beta-22809",
+//    "System.Threading.ThreadPool": "4.0.10-beta-22809"
+  }
+}
```
Comment (stephentoub) [inline src/System.Collections/src/project.json]: Nit: missing newline

File: src/System.Collections/src/project.json (hunk 24 above)
Comment (FiveTimesTheFun) [inline src/System.Collections/src/project.json]: Pardon, I've removed those.

File: src/System.Collections/src/System/Collections/BitArray.cs (hunk 3 above)
Comment (FiveTimesTheFun) [inline src/System.Collections/src/System/Collections/BitArray.cs]: Yep, thanks Stephen.  Updated, and filed https://github.com/dotnet/codeformatter/issues/117 .

File: src/System.Collections/src/System/Collections/Generic/Queue.cs (hunk 2 above)
Comment (FiveTimesTheFun) [inline src/System.Collections/src/System/Collections/Generic/Queue.cs]: That looks accurate.  I checked the rest of the fields and they're used.  I've pushed a commit to remove _ShrinkThreshold.

File: src/System.Collections/src/System/Collections/Generic/List.cs (hunk 19 above)
Comment (FiveTimesTheFun) [inline src/System.Collections/src/System/Collections/Generic/List.cs]: Thanks Stephen - updated.

File: src/System.Collections/src/System/Collections/Generic/Queue.cs (hunk 20 above)
Comment (FiveTimesTheFun) [inline src/System.Collections/src/System/Collections/Generic/Queue.cs]: Thanks Stephen.  Code formatter fixed all the statics, and I manually fixed all the constants.

File: src/System.Collections/src/project.json (hunk 25 above)
Comment (FiveTimesTheFun) [inline src/System.Collections/src/project.json]: Thanks - updated.

File: src/System.Collections/src/System/Collections/Generic/BitHelper.cs (hunk 4 above)
Comment (FiveTimesTheFun) [inline src/System.Collections/src/System/Collections/Generic/BitHelper.cs]: I believe it's because the class has a local field that is a pointer to stack memory.

File: src/System.Collections/src/System/Collections/Generic/Dictionary.cs (hunk 13 above)
Comment (FiveTimesTheFun) [inline src/System.Collections/src/System/Collections/Generic/Dictionary.cs]: We type forward to mscorlib for Dictionary here, but the same comment applies to HashSet.

File: src/System.Collections/src/System/Collections/Generic/BitHelper.cs (hunk 5 above)
Comment (FiveTimesTheFun) [inline src/System.Collections/src/System/Collections/Generic/BitHelper.cs]: Thanks, updated.

File: src/System.Collections/src/System/Collections/Generic/Comparer.cs (hunk 6 above)
Comment (FiveTimesTheFun) [inline src/System.Collections/src/System/Collections/Generic/Comparer.cs]: Comparer is type forwarded to mscorlib, I've updated my initial commit to remove this file.

File: src/System.Collections/src/System/Collections/Generic/Comparer.cs (hunk 7 above)
Comment (FiveTimesTheFun) [inline src/System.Collections/src/System/Collections/Generic/Comparer.cs]: Comparer is type forwarded to mscorlib, I've updated my initial commit to remove this file.

File: src/System.Collections/src/System/Collections/Generic/Dictionary.cs (hunk 11 above)
Comment (FiveTimesTheFun) [inline src/System.Collections/src/System/Collections/Generic/Dictionary.cs]: Dictionary is type forwarded to mscorlib, I've updated my initial commit to remove this file.

File: src/System.Collections/src/System/Collections/Generic/EqualityComparer.cs (hunk 14 above)
Comment (FiveTimesTheFun) [inline src/System.Collections/src/System/Collections/Generic/EqualityComparer.cs]: EqualityComparer is type forwarded to mscorlib, I've updated my initial commit to remove this file.

File: src/System.Collections/src/System/Collections/Generic/EqualityComparer.cs (hunk 15 above)
Comment (FiveTimesTheFun) [inline src/System.Collections/src/System/Collections/Generic/EqualityComparer.cs]: EqualityComparer is type forwarded to mscorlib, I've updated my initial commit to remove this file.

File: src/System.Collections/src/System/Collections/Generic/LinkedList.cs (hunk 16 above)
Comment (FiveTimesTheFun) [inline src/System.Collections/src/System/Collections/Generic/LinkedList.cs]: I believe these were used in GetObjectData and OnDeserialization, support for which was removed from this class.  I'll remove the fields.

File: src/System.Collections/src/System/Collections/Generic/Queue.cs (hunk 20 above)
Comment (FiveTimesTheFun) [inline src/System.Collections/src/System/Collections/Generic/Queue.cs]: Updated.

File: src/System.Collections/src/System/Collections/Generic/SortedSet.cs (hunk 22 above)
Comment (FiveTimesTheFun) [inline src/System.Collections/src/System/Collections/Generic/SortedSet.cs]: Thank you - updated.

File: src/System.Collections/src/System/Collections/Generic/LinkedList.cs (hunk 16 above)
Comment (FiveTimesTheFun) [inline src/System.Collections/src/System/Collections/Generic/LinkedList.cs]: Updated.

File: src/System.Collections/src/System/Collections/Generic/List.cs (hunk 18 above)
Comment (FiveTimesTheFun) [inline src/System.Collections/src/System/Collections/Generic/List.cs]: Thanks - using Array.Empty<T>() in Queue and Stack.

File: src/System.Collections/src/System/Collections/Generic/Dictionary.cs (hunk 13 above)
Comment (FiveTimesTheFun) [inline src/System.Collections/src/System/Collections/Generic/Dictionary.cs]: Filed https://github.com/dotnet/corefx/issues/1534

File: src/System.Collections/src/System/Collections/Generic/LinkedList.cs (hunk 17 above)
Comment (FiveTimesTheFun) [inline src/System.Collections/src/System/Collections/Generic/LinkedList.cs]: There aren't that many others - I was under the impression this was in the same bucket as security transparency attributes - don't add but leave.

File: src/System.Collections/src/System/Collections/Generic/SortedSet.cs (hunk 23 above)
Comment (FiveTimesTheFun) [inline src/System.Collections/src/System/Collections/Generic/SortedSet.cs]: Unfortunately my history is incomplete for when that change was entered in 2008.  All of the methods that use that define were introduced at the same time as the define in my history.  It's currently unset on ProjectK and desktop.  I'll remove this.

Comment (FiveTimesTheFun) [inline src/System.Collections/src/System/Collections/Generic/SortedSet.cs]: Updated.

File: src/System.Collections/src/System/Collections/Generic/QueueDebugView.cs (hunk 21 above)
Comment (FiveTimesTheFun) [inline src/System.Collections/src/System/Collections/Generic/QueueDebugView.cs]: That would be ICollection<T>, right?  We already have a separate debug view for that, ICollectionDebugView<T>, which makes a copy of the contents.

File: src/System.Collections/src/System/Collections/Generic/LinkedList.cs (hunk 17 above)
Comment (FiveTimesTheFun) [inline src/System.Collections/src/System/Collections/Generic/LinkedList.cs]: Apparently the guidance has changed (for security transparency attributes, too).
Updated to remove all the ComVisibleAttributes and Security*Attributes.

File: src/System.Collections/src/System/Collections/Generic/QueueDebugView.cs (hunk 21 above)
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/QueueDebugView.cs]: If the end result is the same, can we just use that then, rather than having a debug view specific to Queue?

File: src/Common/src/System/Collections/Generic/LowLevelComparer.cs (hunk 8 above)
Comment (weshaggard) [inline src/Common/src/System/Collections/Generic/LowLevelComparer.cs]: I have a similar question.

Comment (FiveTimesTheFun) [inline src/Common/src/System/Collections/Generic/LowLevelComparer.cs]: This doesn't appear to be used anymore, I've removed it.

File: src/System.Collections/src/System/Collections/Generic/QueueDebugView.cs (hunk 21 above)
Comment (FiveTimesTheFun) [inline src/System.Collections/src/System/Collections/Generic/QueueDebugView.cs]: Sure - I've just done this for HashSet and SortedSet.  I didn't do this for Queue and Stack, since they don't implement the generic ICollection<> interface.  The extension method for ToArray(this IEnumerable<>) is up in System.Linq.

**Comment by FiveTimesTheFun** [comment]
Thanks Stephen and Wes - I believe I've addressed all the comments and rebased this on top of master.

**Comment by FiveTimesTheFun** [comment]
Thanks @stephentoub and @weshaggard  - I believe I've addressed all the comments and rebased this on top of master.

**Comment by mellinoe** [comment]
You'll want to add the line to the test project file, as Stephen did here:

https://github.com/dotnet/corefx/pull/1541

Other than that, this looks good to me.

File: src/Common/src/System/Collections/HashHelpers.cs [hunk 26]
```diff
@@ -0,0 +1,107 @@
...
+        // Doubling is important for preserving the asymptotic complexity of the 
+        // hashtable operations such as add.  Having a prime guarantees that double 
+        // hashing does not lead to infinite loops.  IE, your hash function will be 
+        // h1(key) + i*h2(key), 0 <= i < size.  h2 and the size must be relatively prime.
+        public static readonly int[] primes = {
+            3, 7, 11, 17, 23, 29, 37, 47, 59, 71, 89, 107, 131, 163, 197, 239, 293, 353, 431, 521, 631, 761, 919,
+            1103, 1327, 1597, 1931, 2333, 2801, 3371, 4049, 4861, 5839, 7013, 8419, 10103, 12143, 14591,
+            17519, 21023, 25229, 30293, 36353, 43627, 52361, 62851, 75431, 90523, 108631, 130363, 156437,
+            187751, 225307, 270371, 324449, 389357, 467237, 560689, 672827, 807403, 968897, 1162687, 1395263,
+            1674319, 2009191, 2411033, 2893249, 3471899, 4166287, 4999559, 5999471, 7199369};
```
Comment (krwq) [inline src/Common/src/System/Collections/HashHelpers.cs]: Nit: it feels like there is a spacebar before } missing. Could be formatter bug

Comment (nydehi) [inline src/Common/src/System/Collections/HashHelpers.cs]: isnt  5 a prime as well?? or is that left out for a reason?

**Comment by FiveTimesTheFun** [comment]
@krwq - The latest guidance from @weshaggard is to remove stale ComVisible attributes, SecurityCritical/SafeCritical/Transparent attributes because those two features are not supported on .NET Core, and while leaving the attributes while porting doesn't harm anything, they become stale and inaccurate or incomplete over time.

File: src/Common/src/System/Collections/HashHelpers.cs (hunk 26 above)
Comment (FiveTimesTheFun) [inline src/Common/src/System/Collections/HashHelpers.cs]: @krwq - updated
@nydehi  - this list of primes is not exhaustive - it's missing many more than 5.  This list is used to choose a hash table size for HashSet.  By having fewer primes in this list we speed up the implementation of GetPrime(int) with a tradeoff of possibly returning a higher prime than would be necessary.

**Comment by FiveTimesTheFun** [comment]
Thanks @mellinoe  - updated.

**Comment by FiveTimesTheFun** [comment]
@dotnet-bot test this please.
